camel/models/open_source_model.py

```diff
@@ -0,0 +1,111 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from types import GeneratorType
+from typing import Any, Dict, List
+
+from camel.messages import OpenAIMessage
+from camel.models import BaseModelBackend
+from camel.typing import ModelType
+from camel.utils import TokenCounterFactory
+
+
+class OpenSourceModel(BaseModelBackend):
+    r"""Class for interace with OpenAI-API-compatible servers running
+    open-source models.
+    """
+
+    def __init__(
+        self,
+        model_type: ModelType,
+        model_config_dict: Dict[str, Any],
+    ) -> None:
+        r"""Constructor for model backends of Open-source models.
+
+        Args:
+            model_type (ModelType): Model for which a backend is created,
+                one of GPT_* series.
+            model_config_dict (Dict[str, Any]): A dictionary that will
+                be fed into openai.ChatCompletion.create().
+        """
+        super().__init__(model_type, model_config_dict)
+
+        import copy
+        self.model_config_dict = copy.copy(model_config_dict)
```

Comment:
Why do you need copy? deep copying stuff hints that you are doing something wrong

Reply:
Ahh.. this is because `model_config_dict` will be input as argument to OpenAI API and extra information like `model_path` and `server_url` will be poped from this dictionary. But I am afraid the change to the inpu dictionary object (or model config object) will result in some unexpected effect (e.g. the dict will be used later after initializing the agent). So I used a copy of the input argument.
Do you think whether this is necessary or not?

Reply:
Please do not mutate variables. They are called variables but they must be as often constants as possible. Re-assemble the dictionary without unwanted keys with dict comprehension.